lib/elixir/lib/calendar/date.ex

```diff
@@ -757,6 +769,75 @@ defmodule Date do
     end
   end
 
+  @doc """
+  Shifts a date by given units. Raises an `ArgumentError` if the
+  given date is not a valid Calendar.ISO date.
+
+  Available shift units are: `:year, :month, :week, :day` and the shift
+  is always applied in that order.
+
+  `Date.shift/2` defines a week as 7 days and a year as 12 months.
+
+  When shifting by month:
+  - it will shift to the current day of a month
+  - if the current day does not exist in a month, it will shift to the last day of a month
+
+  ## Examples
+
+      iex> Date.shift(~D[2016-01-03], month: 2)
+      ~D[2016-03-03]
+      iex> Date.shift(~D[2016-02-29], month: 1)
+      ~D[2016-03-29]
+      iex> Date.shift(~D[2016-01-31], month: 1)
+      ~D[2016-02-29]
+      iex> Date.shift(~D[2016-01-31], year: 4, day: 1)
+      ~D[2020-02-01]
+
+  """
+  # @doc since: "1.5.0"
+  @spec shift(Calendar.date(), [shift_unit()]) :: t
+  def shift(%{calendar: Calendar.ISO} = date, shift_units) do
+    shift_units = Keyword.validate!(shift_units, year: 0, month: 0, week: 0, day: 0)
+
+    ordered_units = [
+      year: shift_units[:year],
+      month: shift_units[:month],
+      week: shift_units[:week],
+      day: shift_units[:day]
+    ]
+
+    Enum.reduce(ordered_units, date, fn
+      {_opt, 0}, new_date -> new_date
+      {:year, value}, new_date -> shift_months(new_date, value * 12)
+      {:month, value}, new_date -> shift_months(new_date, value)
+      {:week, value}, new_date -> Date.add(new_date, value * 7)
+      {:day, value}, new_date -> Date.add(new_date, value)
+    end)
+  end
+
+  def shift(_date, _opts) do
+    raise ArgumentError, "date invalid or not supported"
+  end
+
+  @doc false
+  defp shift_months(date, 0), do: date
+
+  defp shift_months(%Date{year: year, month: month, day: day}, months) do
+    total_months = year * 12 + month + months - 1
+    new_year = Integer.floor_div(total_months, 12)
+
+    new_month =
+      case rem(total_months, 12) + 1 do
+        0 -> 12
+        new_month -> new_month
+      end
+
+    last_day_of_month = :calendar.last_day_of_the_month(abs(new_year), new_month)
```

Comment:
I believe there is a function in Calendar.ISO we could use for this, but I would consider making this function generic so it works with any calendar (by adding a new callback to Calendar.ISO).

Reply:
Supporting the `Calendar` callbacks would be very desirable. Not all calendars have 12 months, or even the same number of months in every year.  And it would be inconsistent now to add a new function that is explicitly not calendar aware.

Reply:
Agreed, @kipcole9. I do have one question though: Date.add and Date.diff are not calendar specific though. Has that been an issue for you? On the other hand, they are at least clear that they are moving along ISO units.

Reply:
The existing callbacks `calendar.months_in_year/1` and `calendar.days_in_month/2` can be used since these functions are already defined in the `Calendar.ISO` behaviour.

Reply:
> Agreed, @kipcole9. I do have one question though: Date.add and Date.diff are not calendar specific though. Has that been an issue for you? On the other hand, they are at least clear that they are moving along ISO units.

I defined a `Cldr.Calendar` behaviour to fill in some of the blanks and defined a `Cldr.Calendar.plus/3` callback to be able to take `:year`, `:quarter`, `:month`,` :week` and `:day` as units for the same purpose as this PR. Although I haven't implemented `Cldr.Calendar.shfit/3` so its great to see this PR.

Reply:
> I was looking into using `Calendar.months_in_year/1` which works well for the month shift as long as the date is not shifted by more months than a year has (unless it's the ISO calendar, then it always works afaik :)). I am afraid for non-gregorian calendars that would require a different approach altogether.

Agreed, it's tricky. Thats why I added a callback (in my case `Cldr.Calendar.plus/3`) specifically for `:month` and `:year`. You could do the same here - delegate the `:year` and `:month` shifting to the `Calendar.ISO` behaviour and then just implement it for that calendar leaving the rest up to other calendar developers (which probably just means me :-)).

Reply:
> I was looking into using Calendar.months_in_year/1 which works well for the month shift as long as the date is not shifted by more months than a year has (unless it's the ISO calendar, then it always works afaik :)).

If we make this support multiple calendars, the whole logic would be inside the calendar. This function would pretty much just validate the input and call Calendar.shift_date/4 (and Calendar.shift_naive_date_time/7 for the others).